File: src/main/resources/src/js/modules/issue/IssueService.js
angular.module('BugTracker.Issue')
	.service('IssueService', ['$http', function($http) {
		'use strict';
		
		this.getIssueById = getIssueById;
		this.getConstants = getConstants;
		this.listAssignableUsers = listAssignableUsers;
		this.assignUserToIssue = assignUserToIssue;
		
		this.createIssue = createIssue;
		this.modifyIssue = modifyIssue;
		
		this.sendComment = sendComment;
		
		function getIssueById(issueId) {
			return $http.get('/issue/' + issueId).then(function(response) {
				return response.data;
			});
		}
		
		function getConstants() {
			return $http.get('/issue/getConstants').then(function(response) {
				return response.data;
			});
		}
		
		function listAssignableUsers(projectId) {
			return $http.get('/issue/listAssignableUsers/' + projectId).then(function(response) {
				return response.data;
			});
		}
		
		function assignUserToIssue(issueId, userId) {
			return $http.post('/issue/assignUserToIssue', {
				issueId: issueId,
				userId: userId
			}).then(function(response) {
				return response.data;
			});
		}
		
		function createIssue(projectId, issue) {
			return $http.post('/issue/saveIssue/' + projectId, createSaveIssueRequest(issue))
			.then(function(response) {
				return response.data;
			});
		}
		
		function modifyIssue(projectId, issue) {
			return $http.put('/issue/saveIssue/' + projectId + '/' + issue.id, createSaveIssueRequest(issue))
			.then(function(response) {
				return response.data;
			});
		}
		
		function sendComment(issueId, message) {
			return $http.post('/issue/sendComment', {
				issueId: issueId,
				message: message
			}).then(function(response) {
				return response.data;
			});
		}
		
		function createSaveIssueRequest(issue) {
			return {
			    name : issue.name,
			    description : issue.description,
			    reproductionSteps : issue.reproductionSteps,
			    version : issue.version,
			    type : issue.type,
			    status : issue.status,
			    priority : issue.priority,
			    severity : issue.severity,
			};
		}
		
	}]);

File: src/main/java/hu/bme/archi/issue/service/IssueService.java
package hu.bme.archi.issue.service;

import java.util.List;

import org.springframework.security.access.prepost.PostFilter;
import org.springframework.security.access.prepost.PreAuthorize;

import hu.bme.archi.issue.bean.AssignUserToIssueRequest;
import hu.bme.archi.issue.bean.Comment;
import hu.bme.archi.issue.bean.ConstantsResponse;
import hu.bme.archi.issue.bean.CreateNewProjectRequest;
import hu.bme.archi.issue.bean.ListIssuesData;
import hu.bme.archi.issue.bean.ListIssuesResponse;
import hu.bme.archi.issue.bean.ListProjectsData;
import hu.bme.archi.issue.bean.SaveIssueRequest;
import hu.bme.archi.issue.bean.SendCommentRequest;
import hu.bme.archi.issue.bean.User;

@PreAuthorize("denyAll()")
public interface IssueService {

	@PreAuthorize("hasAuthority('LOGIN')")
	@PostFilter("hasPermission(filterObject.id, 'hu.bme.archi.issue.domain.Project', 'assigned')")
	List<ListProjectsData> listProjects();

	@PreAuthorize("hasPermission(#projectId, 'hu.bme.archi.issue.domain.Project', 'assigned')")
	ListIssuesResponse listIssues(long projectId);

	@PreAuthorize("hasAuthority('CREATE_PROJECT')")
	void createNewProject(CreateNewProjectRequest createNewProjectRequest);
	
	@PreAuthorize("hasAuthority('CREATE_ISSUE')")
	void createIssue(long projectId,  SaveIssueRequest saveIssueRequest);
	
	@PreAuthorize("hasAuthority('MODIFY_ISSUE')")
	void saveIssue(long projectId, long issueId,  SaveIssueRequest saveIssueRequest);
	
	@PreAuthorize("hasAuthority('DELETE_ISSUE')")
	void deleteIssue(long issueId);	
	
	@PreAuthorize("hasAuthority('COMMENT_ISSUE') && hasPermission(#sendCommentRequest.issueId, 'hu.bme.archi.issue.domain.Issue', 'assigned')")
	List<Comment> sendComment(SendCommentRequest sendCommentRequest);
	
	@PreAuthorize("hasPermission(#issueId, 'hu.bme.archi.issue.domain.Issue', 'assigned')")
	ListIssuesData getIssueById(long issueId);

	@PreAuthorize("hasAuthority('CREATE_ISSUE') || hasAuthority('MODIFY_ISSUE')")
	ConstantsResponse getConstants();
	
	@PreAuthorize("hasPermission(#projectId, 'hu.bme.archi.issue.domain.Project', 'assigned')")
	List<User> listAssignableUsers(long projectId);

	@PreAuthorize("hasAuthority('MODIFY_ISSUE')")
	void assignUserToIssue(AssignUserToIssueRequest assignUserToIssueRequest);
}
